### ISSUE TITLE
Possible bug in Bob.js problem 

### ISSUE THREAD

**Issue by anthonygedeon** [exercism/exercism#5321]
I'm getting two different answers when running the bob.js

**When Running Tests**

```javascript
test('asking a numeric question', () => {
    const result = hey('You are, what, like 15?');
    expect(result).toEqual('Sure.');
  });
```

**Output**

```javascript
// Terminal
Whatever.
```

**Running node**

```javascript
const REGEX = {
	question: /[a-z|\d]+\?/g,
	yell: /[A-Z]{3,}/g,
	questionYell: /[A-Z]{3,}\?/g,
	nothing: /^[\s]/gm, 
};

const hey = (message) => {
	
	let bobSaid = '';

	switch (true) {

		case REGEX.question.test(message):
			bobSaid = 'Sure.';
			break;

		case REGEX.yell.test(message):
			bobSaid = 'Whoa, chill out!';
			break;

		case REGEX.questionYell.test(message):
			bobSaid = "Calm down, I know what I'm doing!";
			break;

		case REGEX.nothing.test(message):
			bobSaid = 'Fine. Be that way!';
			break;

		default:
			bobSaid = 'Whatever.';
	}

	return bobSaid;
};


console.log(hey('You are, what, like 15?')) // This was removed and I exported the function but the problem still occured
```

**output**

```
node bob.js 
Sure.
```

This could be an error on my part but I don't know why I'm getting the correct answer when I run node and the incorrect answer when running the test

**Comment by SleeplessByte** [comment]
The issue is that you've made the Regular Expressions _[global](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/global)_. This means that the regular expression will search for _all instances_ inside a string. The last found index is stored, and that's exactly what's causing the inconsistent behaviour. Here follows an explanation on what's going on.

Without the global flag, the `test` will always return `true` for the following example:
```javascript
const hasABC = /ABC/

hasABC.test("All I want is ABC, yes ABC")
// => true
hasABC.test("All I want is ABC, yes ABC")
// => true
hasABC.test("All I want is ABC, yes ABC")
// => true
hasABC.test("All I want is ABC, yes ABC")
// => true
```

Once we add the global flag, things are different!
```javascript
const hasABCg = /ABC/g

hasABCg.test("All I want is ABC, yes ABC")
// => true
hasABCg.test("All I want is ABC, yes ABC")
// => true
hasABCg.test("All I want is ABC, yes ABC")
// => false
hasABCg.test("All I want is ABC, yes ABC")
// => true
```

As you can see, the third one returns false. `.test` hides _why_ this happens. Let's replace `.test` with [`.exec`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec):

```javascript
hasABC.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 14, input: "All I want is ABC, yes ABC", groups: undefined]
hasABC.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 14, input: "All I want is ABC, yes ABC", groups: undefined]
hasABC.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 14, input: "All I want is ABC, yes ABC", groups: undefined]
hasABC.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 14, input: "All I want is ABC, yes ABC", groups: undefined]
```

Great. It's matching the first ABC, before the comma, each time. Works! Now let's try the global variant:

```javascript
hasABCg.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 14, input: "All I want is ABC, yes ABC", groups: undefined]
hasABCg.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 23, input: "All I want is ABC, yes ABC", groups: undefined]
hasABCg.exec("All I want is ABC, yes ABC")
// => null
hasABCg.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 14, input: "All I want is ABC, yes ABC", groups: undefined]
```

As you can see, _each_ execution returns the next match in the string, until there are no more matches, which is when it returns `null`. Then, it starts at the beginning again.

You might ask yourself: But I put in a fresh `string` each time?!. That's right. But the state is stored inside the Regular Expression (`RegExp`), not the `string`. 

```javascript
hasABCg.exec("All I want is ABC, yes ABC")
// => ["ABC", index: 14, input: "All I want is ABC, yes ABC", groups: undefined]

hasABCg.exec("ABC should work")
// => null
```

Here, the regular expression says it can't match `ABC` in the second string, because the internal pointer was already at index 17. And because you're using state-full regular expressions, running the tests is _different_ than running the code in node, because [in the tests](https://github.com/exercism/javascript/blob/master/exercises/bob/bob.spec.js), there is state-full context:

```javascript
  test('asking a question', () => {
    const result = hey('Does this cryogenic chamber make me look fat?');
    expect(result).toEqual('Sure.');
  });

  test('asking a numeric question', () => {
    const result = hey('You are, what, like 15?');
    expect(result).toEqual('Sure.');
  });
```

On Exercism, we execute the tests in order (top to bottom), so this is what happens:

```javascript
const question = /[a-z|\d]+\?/g

question.exec('Does this cryogenic chamber make me look fat?')
// => ["fat?", index: 41, input: "Does this cryogenic chamber make me look fat?", groups: undefined]
question .exec('You are, what, like 15?')
// => null
```

## How to solve?

In this case, we don't need the global flag. You don't need to perform multiple matches. In fact I would assert that you don't even need a regular expression, but alas. If you remove the global flag, the returned value is:

```javascript
// => ["15?", index: 20, input: "You are, what, like 15?", groups: undefined]
```

You wouldn't have noticed it if your Regular Expressions were inlined, but that just hides the fact that this is happening in the first place:

```javascript
/[a-z|\d]+\?/g.text(input)
// => will always be a NEW regular expression, so would work
```

## TL;DR

> JavaScript `RegExp` objects are stateful when they have the global or sticky flags set (e.g. `/foo/g` or `/foo/y`). They store a `lastIndex` from the previous match. Using this internally, `exec()` can be used to iterate over multiple matches in a string of text (with capture groups), as opposed to getting just the matching strings with `String.prototype.match()`.

This means that calling `.test`, which uses `.exec` under the hood has a side-effect of moving the internal pointer where it starts reading the string. This causes the second test with a question mark to fail (as it returns `false`, instead of the expected `true`).

**Comment by ErikSchierboom** [comment]
Hello 👋

With the launch of Exercism v3, we are **closing all issues in this repository** to help give us a clean slate to detect new problems. If this issue is still relevant to Exercism v3 (e.g. it's a feature that we haven't implemented in v3, or a bug that still exists), please reopen it and we will review it and post an update on it as soon as we get chance.

Thanks for helping make Exercism better, and we hope you enjoy v3 🙂
